#      **Project 2.-San Francisco City Employee Salary  DataSet Analysis**
### **Introduction**
      
The SF Salaries dataset analysis delves into the compensation structures of San Francisco city employees, focusing on various aspects of their pay and benefits. Through detailed examination, the project explores trends, identifies key patterns, and provides insights into employee remuneration. This analysis covers crucial data points such as base pay, overtime, benefits, and job titles, offering a comprehensive view of the city's payroll.
dataset taken from this (https://www.kaggle.com/datasets/kaggle/sf-salaries).

In [ ]:
#Import Liabraries of Python
import pandas as pd
#Load Dataset
df=pd.read_csv('Salaries.csv')

<ipython-input-292-85880d0ec06a>:2: DtypeWarning: Columns (3,4,5,6,12) have mixed types. Specify dtype option on import or set low_memory=False.
  df=pd.read_csv('Salaries.csv')


In [ ]:
# Display Top 15 Rows of The Dataset
df.head(15)

Id        EmployeeName                                          JobTitle  \
0    1      NATHANIEL FORD    GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY   
1    2        GARY JIMENEZ                   CAPTAIN III (POLICE DEPARTMENT)   
2    3      ALBERT PARDINI                   CAPTAIN III (POLICE DEPARTMENT)   
3    4   CHRISTOPHER CHONG              WIRE ROPE CABLE MAINTENANCE MECHANIC   
4    5     PATRICK GARDNER      DEPUTY CHIEF OF DEPARTMENT,(FIRE DEPARTMENT)   
5    6      DAVID SULLIVAN                         ASSISTANT DEPUTY CHIEF II   
6    7           ALSON LEE                BATTALION CHIEF, (FIRE DEPARTMENT)   
7    8       DAVID KUSHNER                    DEPUTY DIRECTOR OF INVESTMENTS   
8    9      MICHAEL MORRIS                BATTALION CHIEF, (FIRE DEPARTMENT)   
9   10  JOANNE HAYES-WHITE            CHIEF OF DEPARTMENT, (FIRE DEPARTMENT)   
10  11       ARTHUR KENNEY  ASSISTANT CHIEF OF DEPARTMENT, (FIRE DEPARTMENT)   
11  12    PATRICIA JACKSON                   CAPTAIN III (POLICE DEPARTMENT)   
12  13   EDWARD HARRINGTON                       EXECUTIVE CONTRACT EMPLOYEE   
13  14         JOHN MARTIN                                 DEPARTMENT HEAD V   
14  15      DAVID FRANKLIN                BATTALION CHIEF, (FIRE DEPARTMENT)   

      BasePay OvertimePay   OtherPay Benefits   TotalPay  TotalPayBenefits  \
0   167411.18         0.0  400184.25      NaN  567595.43         567595.43   
1   155966.02   245131.88  137811.38      NaN  538909.28         538909.28   
2   212739.13   106088.18    16452.6      NaN  335279.91         335279.91   
3     77916.0    56120.71   198306.9      NaN  332343.61         332343.61   
4    134401.6      9737.0  182234.59      NaN  326373.19         326373.19   
5    118602.0      8601.0  189082.74      NaN  316285.74         316285.74   
6    92492.01     89062.9  134426.14      NaN  315981.05         315981.05   
7   256576.96         0.0    51322.5      NaN  307899.46         307899.46   
8   176932.64    86362.68   40132.23      NaN  303427.55         303427.55   
9    285262.0         0.0   17115.73      NaN  302377.73         302377.73   
10  194999.39    71344.88    33149.9      NaN  299494.17         299494.17   
11    99722.0    87082.62   110804.3      NaN  297608.92         297608.92   
12  294580.02         0.0        0.0      NaN  294580.02         294580.02   
13  271329.03         0.0   21342.59      NaN  292671.62         292671.62   
14  174872.64     74050.3   37424.11      NaN  286347.05         286347.05   

    Year  Notes         Agency Status  
0   2011    NaN  San Francisco    NaN  
1   2011    NaN  San Francisco    NaN  
2   2011    NaN  San Francisco    NaN  
3   2011    NaN  San Francisco    NaN  
4   2011    NaN  San Francisco    NaN  
5   2011    NaN  San Francisco    NaN  
6   2011    NaN  San Francisco    NaN  
7   2011    NaN  San Francisco    NaN  
8   2011    NaN  San Francisco    NaN  
9   2011    NaN  San Francisco    NaN  
10  2011    NaN  San Francisco    NaN  
11  2011    NaN  San Francisco    NaN  
12  2011    NaN  San Francisco    NaN  
13  2011    NaN  San Francisco    NaN  
14  2011    NaN  San Francisco    NaN

In [ ]:
#Display Last 5 Rows of The Dataset
df.tail(5)

Id   EmployeeName                    JobTitle       BasePay  \
148649  148650  Roy I Tillery                   Custodian          0.00   
148650  148651   Not provided                Not provided  Not Provided   
148651  148652   Not provided                Not provided  Not Provided   
148652  148653   Not provided                Not provided  Not Provided   
148653  148654      Joe Lopez  Counselor, Log Cabin Ranch          0.00   

         OvertimePay      OtherPay      Benefits  TotalPay  TotalPayBenefits  \
148649          0.00          0.00          0.00      0.00              0.00   
148650  Not Provided  Not Provided  Not Provided      0.00              0.00   
148651  Not Provided  Not Provided  Not Provided      0.00              0.00   
148652  Not Provided  Not Provided  Not Provided      0.00              0.00   
148653          0.00       -618.13          0.00   -618.13           -618.13   

        Year  Notes         Agency Status  
148649  2014    NaN  San Francisco     PT  
148650  2014    NaN  San Francisco    NaN  
148651  2014    NaN  San Francisco    NaN  
148652  2014    NaN  San Francisco    NaN  
148653  2014    NaN  San Francisco     PT

In [ ]:
# Find Shape of Our Dataset (Number of Rows And Number of Columns)
df.shape
print("Number of Rows",df.shape[0])
print("Number of Columns",df.shape[1])

Number of Rows 148654
Number of Columns 13


In [ ]:
# Getting Information About Our Dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 148654 entries, 0 to 148653
Data columns (total 13 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   Id                148654 non-null  int64  
 1   EmployeeName      148654 non-null  object 
 2   JobTitle          148654 non-null  object 
 3   BasePay           148049 non-null  object 
 4   OvertimePay       148654 non-null  object 
 5   OtherPay          148654 non-null  object 
 6   Benefits          112495 non-null  object 
 7   TotalPay          148654 non-null  float64
 8   TotalPayBenefits  148654 non-null  float64
 9   Year              148654 non-null  int64  
 10  Notes             0 non-null       float64
 11  Agency            148654 non-null  object 
 12  Status            38119 non-null   object 
dtypes: float64(3), int64(2), object(8)
memory usage: 14.7+ MB


In [ ]:
# Convert Datatype of columns Basepay,Overtimepay,otherpay from object to float
df['BasePay']=pd.to_numeric(df['BasePay'],errors='coerce')
df['OvertimePay']=pd.to_numeric(df['OvertimePay'],errors='coerce')
df['OtherPay']=pd.to_numeric(df['OtherPay'],errors='coerce')

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 148654 entries, 0 to 148653
Data columns (total 13 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   Id                148654 non-null  int64  
 1   EmployeeName      148654 non-null  object 
 2   JobTitle          148654 non-null  object 
 3   BasePay           148045 non-null  float64
 4   OvertimePay       148650 non-null  float64
 5   OtherPay          148650 non-null  float64
 6   Benefits          112495 non-null  object 
 7   TotalPay          148654 non-null  float64
 8   TotalPayBenefits  148654 non-null  float64
 9   Year              148654 non-null  int64  
 10  Notes             0 non-null       float64
 11  Agency            148654 non-null  object 
 12  Status            38119 non-null   object 
dtypes: float64(6), int64(2), object(5)
memory usage: 14.7+ MB


In [ ]:
#Check what are columns in the dataset
df.columns

Index(['Id', 'EmployeeName', 'JobTitle', 'BasePay', 'OvertimePay', 'OtherPay',
       'Benefits', 'TotalPay', 'TotalPayBenefits', 'Year', 'Notes', 'Agency',
       'Status'],
      dtype='object')

In [ ]:
# Check any Missing Values in the dataset
df.isnull().sum()

Id                       0
EmployeeName             0
JobTitle                 0
BasePay                609
OvertimePay              4
OtherPay                 4
Benefits             36159
TotalPay                 0
TotalPayBenefits         0
Year                     0
Notes               148654
Agency                   0
Status              110535
dtype: int64

In [ ]:
# check if there is any duplicates
df.duplicated().sum()

0

In [ ]:
#Drop ID, Notes, Agency, and Status Columns
df.columns

Index(['Id', 'EmployeeName', 'JobTitle', 'BasePay', 'OvertimePay', 'OtherPay',
       'Benefits', 'TotalPay', 'TotalPayBenefits', 'Year', 'Notes', 'Agency',
       'Status'],
      dtype='object')

In [ ]:
# Drop ID, Notes, Agency, and Status Columns without changing original dataset
df.drop(['Id','Agency','Notes','Status'],axis=1,inplace=True)
df.head(1)

EmployeeName                                        JobTitle    BasePay  \
0  NATHANIEL FORD  GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY  167411.18   

   OvertimePay   OtherPay Benefits   TotalPay  TotalPayBenefits  Year  
0          0.0  400184.25      NaN  567595.43         567595.43  2011

In [ ]:
df.columns

Index(['EmployeeName', 'JobTitle', 'BasePay', 'OvertimePay', 'OtherPay',
       'Benefits', 'TotalPay', 'TotalPayBenefits', 'Year'],
      dtype='object')

In [ ]:
# Get Overall Statistics About The Dataframe
df.describe(include='all')

EmployeeName          JobTitle        BasePay    OvertimePay  \
count        148654            148654  148045.000000  148650.000000   
unique       110811              2159            NaN            NaN   
top       Kevin Lee  Transit Operator            NaN            NaN   
freq             13              7036            NaN            NaN   
mean            NaN               NaN   66325.448840    5066.059886   
std             NaN               NaN   42764.635495   11454.380559   
min             NaN               NaN    -166.010000      -0.010000   
25%             NaN               NaN   33588.200000       0.000000   
50%             NaN               NaN   65007.450000       0.000000   
75%             NaN               NaN   94691.050000    4658.175000   
max             NaN               NaN  319275.010000  245131.880000   

             OtherPay  Benefits       TotalPay  TotalPayBenefits  \
count   148650.000000  112495.0  148654.000000     148654.000000   
unique            NaN   99635.0            NaN               NaN   
top               NaN       0.0            NaN               NaN   
freq              NaN    1053.0            NaN               NaN   
mean      3648.767297       NaN   74768.321972      93692.554811   
std       8056.601866       NaN   50517.005274      62793.533483   
min      -7058.590000       NaN    -618.130000       -618.130000   
25%          0.000000       NaN   36168.995000      44065.650000   
50%        811.270000       NaN   71426.610000      92404.090000   
75%       4236.065000       NaN  105839.135000     132876.450000   
max     400184.250000       NaN  567595.430000     567595.430000   

                 Year  
count   148654.000000  
unique            NaN  
top               NaN  
freq              NaN  
mean      2012.522643  
std          1.117538  
min       2011.000000  
25%       2012.000000  
50%       2013.000000  
75%       2014.000000  
max       2014.000000

In [ ]:
# Find Occurrence of The Employee Names  (Top 5)
df['EmployeeName'].value_counts().head(5)

EmployeeName
Kevin Lee       13
Richard Lee     11
Steven Lee      11
William Wong    11
Stanley Lee      9
Name: count, dtype: int64

In [ ]:
# Find The Number of Unique Job Titles
df['JobTitle'].nunique()


2159

In [ ]:
# Total Number of Job Titles Contain Captain
len(df[df['JobTitle'].str.contains('Captain',case=False)])

552

In [ ]:
df[df['JobTitle'].str.contains('Captain',case=False)].count()

EmployeeName        552
JobTitle            552
BasePay             551
OvertimePay         552
OtherPay            552
Benefits            411
TotalPay            552
TotalPayBenefits    552
Year                552
dtype: int64

In [ ]:
# Display All the Employee Names From Fire Department
df[df['JobTitle'].str.contains('Fire',case=False)]['EmployeeName']

4            PATRICK GARDNER
6                  ALSON LEE
8             MICHAEL MORRIS
9         JOANNE HAYES-WHITE
10             ARTHUR KENNEY
                 ...        
145956      Kenneth C Farris
147556         Edward A Dunn
148021        Kari A Johnson
148209          Sheryl K Lee
148554       Lawrence F Gatt
Name: EmployeeName, Length: 5879, dtype: object

In [ ]:
# Find Minimum, Maximum, and Average BasePay
df['BasePay'].describe()


count    148045.000000
mean      66325.448840
std       42764.635495
min        -166.010000
25%       33588.200000
50%       65007.450000
75%       94691.050000
max      319275.010000
Name: BasePay, dtype: float64

In [ ]:
# Replace 'Not Provided' in EmployeeName' Column to NaN
df['EmployeeName']

0            NATHANIEL FORD
1              GARY JIMENEZ
2            ALBERT PARDINI
3         CHRISTOPHER CHONG
4           PATRICK GARDNER
                ...        
148649        Roy I Tillery
148650         Not provided
148651         Not provided
148652         Not provided
148653            Joe Lopez
Name: EmployeeName, Length: 148654, dtype: object

In [ ]:
import numpy as np
df['EmployeeName'].replace('Not provided',np.nan,inplace=True)

In [ ]:
df['EmployeeName']

0            NATHANIEL FORD
1              GARY JIMENEZ
2            ALBERT PARDINI
3         CHRISTOPHER CHONG
4           PATRICK GARDNER
                ...        
148649        Roy I Tillery
148650                  NaN
148651                  NaN
148652                  NaN
148653            Joe Lopez
Name: EmployeeName, Length: 148654, dtype: object

In [ ]:
# Find The Rows Having 5 Missing Values
df['Benefits'].replace('Not Provided',np.nan,inplace=True)
df[df.isnull().sum(axis=1)==5]

EmployeeName      JobTitle  BasePay  OvertimePay  OtherPay Benefits  \
148646          NaN  Not provided      NaN          NaN       NaN      NaN   
148650          NaN  Not provided      NaN          NaN       NaN      NaN   
148651          NaN  Not provided      NaN          NaN       NaN      NaN   
148652          NaN  Not provided      NaN          NaN       NaN      NaN   

        TotalPay  TotalPayBenefits  Year  
148646       0.0               0.0  2014  
148650       0.0               0.0  2014  
148651       0.0               0.0  2014  
148652       0.0               0.0  2014

In [ ]:
# Drop the rows where 5 values are missing or nUll values
df.isnull().sum(axis=1)

0         1
1         1
2         1
3         1
4         1
         ..
148649    0
148650    5
148651    5
148652    5
148653    0
Length: 148654, dtype: int64

In [ ]:
# Drop rows where 5 values are missing or nUll values
df.drop(df[df.isnull().sum(axis=1)==5].index,axis=0,inplace=True)

In [ ]:
df.isnull().sum(axis=1)

0         1
1         1
2         1
3         1
4         1
         ..
148645    0
148647    0
148648    0
148649    0
148653    0
Length: 148650, dtype: int64

In [ ]:
# Find Job Title of ALBERT PARDINI
df[df['EmployeeName']=='ALBERT PARDINI']['JobTitle']

2    CAPTAIN III (POLICE DEPARTMENT)
Name: JobTitle, dtype: object

In [ ]:
# How Much ALBERT PARDINI Make (Include Benefits)?
df[df['EmployeeName']=='ALBERT PARDINI']['TotalPayBenefits']


2    335279.91
Name: TotalPayBenefits, dtype: float64

In [ ]:
# Display Name of The Person Having The Highest BasePay
df[df['BasePay']==df['BasePay'].max()]['EmployeeName']

72925    Gregory P Suhr
Name: EmployeeName, dtype: object

In [ ]:
# Find Average BasePay of All Employee Per Year
df['BasePay'].groupby(df['Year']).mean()

Year
2011    63595.956517
2012    65436.406857
2013    69630.030216
2014    66564.421924
Name: BasePay, dtype: float64

In [ ]:
# Find Average BasePay of All Employee Per JobTitle
df.columns

Index(['EmployeeName', 'JobTitle', 'BasePay', 'OvertimePay', 'OtherPay',
       'Benefits', 'TotalPay', 'TotalPayBenefits', 'Year'],
      dtype='object')

In [ ]:
df['BasePay'].groupby(df['JobTitle']).mean()


JobTitle
ACCOUNT CLERK                                     43300.806506
ACCOUNTANT                                        46643.172000
ACCOUNTANT INTERN                                 28732.663958
ACPO,JuvP, Juv Prob (SFERS)                       62290.780000
ACUPUNCTURIST                                     66374.400000
                                                      ...     
X-RAY LABORATORY AIDE                             47664.773077
X-Ray Laboratory Aide                             46086.387100
YOUTH COMMISSION ADVISOR, BOARD OF SUPERVISORS    52609.910000
Youth Comm Advisor                                39077.957500
ZOO CURATOR                                       43148.000000
Name: BasePay, Length: 2158, dtype: float64

In [ ]:
# Find Average BasePay of Employee Having Job Title ACCOUNTANT
df[df['JobTitle']=='ACCOUNTANT']['BasePay'].mean()

46643.172

In [ ]:
 # Find Top 5 Most Common Jobs
df['JobTitle'].value_counts()[:5]

JobTitle
Transit Operator                7036
Special Nurse                   4389
Registered Nurse                3736
Public Svc Aide-Public Works    2518
Police Officer 3                2421
Name: count, dtype: int64

In [357]:
# How does overtime pay vary by job title?
df.groupby('JobTitle')['OvertimePay'].mean()

JobTitle
ACCOUNT CLERK                                      373.200843
ACCOUNTANT                                           0.000000
ACCOUNTANT INTERN                                   24.430625
ACPO,JuvP, Juv Prob (SFERS)                          0.000000
ACUPUNCTURIST                                        0.000000
                                                     ...     
X-RAY LABORATORY AIDE                             3571.223462
X-Ray Laboratory Aide                             3483.767100
YOUTH COMMISSION ADVISOR, BOARD OF SUPERVISORS       0.000000
Youth Comm Advisor                                   0.000000
ZOO CURATOR                                          0.000000
Name: OvertimePay, Length: 2158, dtype: float64

# **Conclusion:**

**Based on the analysis of the SF Salaries dataset:**



*   Base pay shows a steady increase from 2011 to 2013, with a slight dip in 2014.
*   Employees in roles like Transit Operator and Special Nurse are most prevalent.
*   Accountant job titles generally command an average base pay of approximately $46,643.
*   Overtime pay varies significantly across job titles, with some roles like X-Ray.
*   Laboratory Aide showing substantial overtime earnings.



**Overall, this analysis provides insights into salary trends and job distribution among San Francisco city employees, highlighting areas for further investigation into compensation disparities and workload distribution.**







